### PR TITLE
Support directly passing a cppexpose::Object to a script function

### PR DESCRIPTION
**Pull request by j-o**
As mentioned in #51.

This PR contains some changes that might be worth discussing:
- Each object wrapper (even for sub-objects) is now created and owned through the backend
- Object wrappers are retained until the wrapped cppexpose Object is destroyed (or the backend)
- Consequently, calling `removeGlobalObject` only removes the key from the global script object, but does not delete the wrapper
- This ensures that there is (at most) a single unique script wrapper object for each cppexpose Object, and thus adding an object as global object and later passing the same object directly to a script function actually refers to the same script object
- There is on small caveat: the Variant used to pass the Object must be created with `cppexpose::Object *` as internal value type (i.e., `Variant::fromValue<cppexpose::Object *>(myObject)`), regardless of the actual subclass.

Is it desirable to implement the other direction as well? Currently, if the duktape backend encounters an object, it creates a VariantMap of all its properties. Since the return value of `fromDukStack` is a Variant, it would theoretically be possible to return either that map, or a pointer to the cppexpose Object, depending on whether the Object has been registered before.

Here is a small test snippet that demonstrates this using the script-example (cf. scripting-example/main.cpp:98):
```
> script.obj.setFunction(function(object) { print(object.string);});
Setting function.
> script.obj.callFunction();
Hallo
> script.obj.string = "foobar";
foobar
> script.obj.callFunction();
foobar
```

### CODE REVIEW

**Comment by sbusch42** [comment]
Very nice!

Can you also implement the other direction in fromDukStack? If the object is a wrapper to a cppexpose::Object, extract the pointer and return it as a Variant. Otherwise, create a VariantMap as before. Then we can pass cppexpose::Object pointers around in scripting and return them to C++ at any point, which will make the scripting environment more or less complete :)

**Comment by j-o** [comment]
Implemented this in fromDukStack

**Comment by sbusch42** [comment]
Added support for using cppexpose::Object * parameters in exposed functions like this:

```
MyObject() {
  addFunction("doWithObject", this, &MyObject::doWithObject);
}

void MyObject::doWithObject(cppexpose::Object * obj) {
  ...
}
```

File: source/cppexpose/source/scripting/DuktapeScriptBackend.cpp [hunk 1]
```diff
@@ -210,22 +203,33 @@ Variant DuktapeScriptBackend::fromDukStack(duk_idx_t index)
...
 
         duk_enum(m_context, index, 0);
         while (duk_next(m_context, -1, 1))
         {
-            // Prevent the pointer to the C++ object that is stored in the Ecmascript object from being serialized
-            if (!(duk_is_pointer(m_context, -1) && fromDukStack(-2).value<std::string>() == s_duktapeObjectPointerKey))
+            // If a property s_duktapeObjectPointerKey exists, the object is a cppexpose::Object.
+            // In this case, extract the pointer and return that below
+            // Otherwise, continue to build a key-value map of the object's properties
+            if (duk_is_pointer(m_context, -1) && fromDukStack(-2).value<std::string>() == s_duktapeObjectPointerKey)
```
Comment (sbusch42) [inline source/cppexpose/source/scripting/DuktapeScriptBackend.cpp]: Can we avoid the while loop altogether here (analog to the Qt implementation in gloperate)?

**Comment by sbusch42** [comment]
Please review my changes, then we can merge.

File: source/cppexpose/source/scripting/DuktapeScriptBackend.cpp (hunk 1 above)
Comment (j-o) [reply source/cppexpose/source/scripting/DuktapeScriptBackend.cpp]: I added an alternative implementation.

**Comment by j-o** [comment]
Looks good.

**Pull request merged**